app/controllers/paymentsAndCharges/PaymentsAndChargeDetailsController.scala

```diff
@@ -61,39 +64,62 @@ class PaymentsAndChargeDetailsController @Inject()(override val messagesApi: Mes
       ) flatMap {
         schemeDetails =>
           fsConnector.getSchemeFS(schemeDetails.pstr).flatMap {
-            seqSchemeFS =>
-              val chargeRefs: Seq[String] =
+            schemeFS =>
+              val schemeFSGroupedAndSorted: Seq[(LocalDate, Seq[SchemeFS])] =
                 paymentsAndChargesService
-                  .groupAndSortByStartDate(seqSchemeFS, startDate.getYear)
-                  .flatMap(_._2)
-                  .map(_.chargeReference)
-
-              try {
-                seqSchemeFS.find(_.chargeReference == chargeRefs(index.toInt)) match {
-                  case Some(schemeFs) =>
-                    renderer.render(
-                      "paymentsAndCharges/paymentsAndChargeDetails.njk",
-                      summaryListData(srn, startDate, schemeFs, schemeDetails.schemeName, index)
-                    )
-                      .map(Ok(_))
-                  case _ =>
-                    Logger.warn(s"No Payments and Charge details found for the selected charge reference ${chargeRefs(index.toInt)}")
-                    Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
-                }
-              } catch {
-                case _: IndexOutOfBoundsException =>
-                  Logger.warn(
-                    s"[paymentsAndCharges.PaymentsAndChargeDetailsController][IndexOutOfBoundsException]:" +
-                      s"index $index of collection length ${chargeRefs.length} attempted"
-                  )
+                  .groupAndSortByStartDate(schemeFS, startDate.getYear)
+
+              val chargeRefsGroupedAndSorted: Seq[(LocalDate, Seq[String])] =
+                schemeFSGroupedAndSorted.map(
+                  dateAndFs => {
+                    val (date, schemeFs) = dateAndFs
+                    (date, schemeFs.map(_.chargeReference))
+                  }
+                )
+
+              (
+                schemeFSGroupedAndSorted.find(_._1 == startDate),
+                chargeRefsGroupedAndSorted.find(_._1 == startDate)
+              ) match {
+                case (Some(dateSchemeFs), Some(dateChargeRefs)) =>
+                  val (_, seqSchemeFs) = dateSchemeFs
```

Comment:
This looks a bit odd. It might be cleaner to just match like this on previous line : case (Some(Tuple2(_, seqSchemeFS)...

Reply:
Done